modules/dataverse-spi/src/main/java/io/gdcc/spi/export/Exporter.java

```diff
@@ -0,0 +1,122 @@
+package io.gdcc.spi.export;
+
+import java.io.OutputStream;
+import java.util.Locale;
+import java.util.Optional;
+
+
+/**
+ * Dataverse allows new metadata export formats to be dynamically added a running instance. This is done by
+ * deploying new classes that implement this Exporter interface.
+ */
+
+public interface Exporter {
+
+
+    /**
+     * When this method is called, the Exporter should write the metadata to the given OutputStream.
+     * 
+     * @apiNote When implementing exportDataset, when done writing content, please make sure
+     * to flush() the outputStream, but NOT close() it! This way an exporter can be
+     * used to insert the produced metadata into the body of an HTTP response, etc.
+     * (for example, to insert it into the body of an OAI response, where more XML
+     * needs to be written, for the outer OAI-PMH record). -- L.A. 4.5
+     * 
+     * @param dataProvider - the @see ExportDataProvider interface includes several methods that can be used to retrieve the dataset metadata in different formats. An Exporter should use one or more of these to obtain the values needed to generate metadata in the format it supports. 
+     * @param outputStream - the OutputStream to write the metadata to
+     * @throws ExportException - if there is an error writing the metadata
+     */
+    void exportDataset(ExportDataProvider dataProvider, OutputStream outputStream) throws ExportException;
```

Comment:
I would not pass an `exportDataProvider` as a parameter to `exportDataset`. First, since the name of the interface and method is `exportDataset`, I assume that I can pass a DataSet to this method to get an external representation. Secondly, to create an export, an implementer may need access to the real typed object and all associated properties. These cannot be contained in a single ExportDataProvider response. Third, ExportDataProvider responses are either JSON or XML objects, and an implementer must parse these even if a typed object tree is available. This leads to unnecessary complexity. Fourth, the responses of `ExportDataProvider` are already results of exports that can potentially be overridden. Therefore, I would change the definition to `void exportDataset(DataSet dataset, OutputStream outputStream) throws ExportException;`

Reply:
This will expose the internal structure to an Exporter. IMHO an Exporter is a piece of code that transfers an internal representation to an external one, subsequently it needs access to the internal structure.

Reply:
Thank you @johannes-darms for the feedback. Let me add some more context you can't possibly know here, that hopefully makes the design choices easier to follow.

1. This work has been funded by DANS, and funding is limited. The changes you are asking for involve the creation of data transfer objects and code translating between the internal database models (`DvObject, Dataset, DataFile, Dataverse`) and those. This task can't be covered on current funding.
2. @qqmyers is aware this is not the technically best solution and kind of introduces technical debt. But it is a starting point to something very important, that we have been longing for for a long time: real modularity with loadable JARs and externalized Java code. I do agree this solution here is limited and far from perfect. "Aber besser den Spatz in der Hand, als die Taube auf dem Dach."

Please keep on nagging to go for the best. Funding and future PRs improving this are very welcome (but we'd need to coordinate such activities).

Reply:
Yes - thanks for the comments! There are definitely aspects of trying to do an incremental update without rewriting the whole export system. 

One thing I'd like to clarify is that the entries in the ExportDataProvider interface are not technically the content of the built-in exporters - they are internal representations of the dataset used in multiple places (to send to DataCite, used in archival Bags, embedded in the header of dataset pages) where a metadata exporter with the same/very similar name has been created to make those formats available for download/use in harvesting. Even before this PR, most of the internal Exporters either parsed the JSON format or just used the DatasetVersion object just to get the DataCiteXML or schema.org output directly. (The two exceptions are the HTML exporter which used the DDI export directly, and the one DDI exporter that provides the detailed variable-level file metadata - it did scan through all the related classes rather than starting from an XML/JSON serialization.) 

Some consequences of all of this are that replacing an internal exporter does not change what goes to DataCite, etc. (True before and after this PR). That would be useful future work - to reverse the logic and have the exporter output used internally rather than vice versa, at which point some of the get* methods could be dropped. Right now the JSON or ORE representations, along with the FileDetails are the ones intended to be complete, so they would be the only ones that would stay or be replaced by something cleaner. 

I would also say that if there is dataset metadata not available to Exporters from the existing interface methods, it's a bug/limitation that should be fixed. With the ORE representation in particular, the intent is to be complete enough to allow recreation of a dataset from an archival Bag, so we want to add anything missing to it for that purpose as well.